mitosheet/src/api/api.tsx

```diff
@@ -75,273 +64,162 @@ export enum UserJsonFields {
-
-        } else if (response['event'] == 'edit_error') {
-            // If the backend sets the data field of the error, then we know
-            // that this is an error that we want to only pass through, without 
-            // displaying an error modal
-            if (response['data'] === undefined) {
-                this.setUIState((prevUIState) => {
-                    return {
-                        ...prevUIState,
-                        currOpenModal: {
-                            type: ModalEnum.Error,
-                            error: (response as unknown) as MitoError
-                        }
-                    }
-                })
             }
-        }
+            return {
+                ...prevUIState,
+                loading: newLoadingCalls
+            }
+        });
+        
     }
+    
+    async send<ResultType>(msg: Record<string, unknown>): Promise<MitoAPIResult<ResultType>> {
 
-    /*
-        Helper function that tries to get the response for a given ID, and returns
-        the data inside the 'data' key in this response if it exists. 
+        // Generate a random id, and add it to the params
+        const id = getRandomId();
+        msg['id'] = id;
 
-        The type of the data that is returned should be the same as the Type given
-        for this generic function.
+        if (this._send === undefined) {
+            const _send = await this.getSendFunction();
+            this._send = this._send || _send;
+        } 
 
-        Returns undefined if it does not get a response within the set timeframe
-        for retries.
-    */
-    getResponseData<Type>(id: string, maxRetries = MAX_RETRIES): Promise<Type | undefined> {
-
-        return new Promise((resolve) => {
-            let tries = 0;
-            const interval = setInterval(() => {
-                // Only try at most MAX_RETRIES times
-                tries++;
-                if (tries > maxRetries) {
-                    console.error(`No response on message: {id: ${id}}`);
-                    clearInterval(interval);
-                    // If we fail, we return an empty response
-                    return resolve(undefined)
-                }
+        await waitUntilConditionReturnsTrueOrTimeout(() => {return this._send !== undefined}, MAX_WEIGHT_FOR_SEND_CREATION);
 
-                // See if there is an API response to this one specificially
-                const index = this.unconsumedResponses.findIndex((response) => {
-                    return response['id'] === id;
-                })
-                if (index !== -1) {
-                    // Clear the interval
-                    clearInterval(interval);
+        if (this._send === undefined) {
+            console.error(`Unable to establish comm. Quitting before sending message with id ${id}`);
+            return {error: 'Connection error. Unable to establish comm.', errorShort: 'Connection error', showErrorModal: true};
+        }
```

Comment:
The reason it is programmed this way is because of race conditions. Namely, multiple calls might go through this function at once, all before the send function is defined. 

In fairness, you're right that (1) is unnecessary. So removing. (2) is necessary because of race conditions so we don't want to overwrite an existing send function!